### ISSUE TITLE
Support for `opam admin make` mirroring with repos containing git URLs

### ISSUE THREAD

**Issue by hcarty** [ocaml/opam#2901]
I have a private opam repository with projects pulled in via git tags along the lines of:
```
git: "git@someserver:user/project.git#v1.0.0
```
It would be very useful if `opam admin make` could mirror archives of these source repo checkouts.  This could help with offline use of source repositories which require authentication to access.

**Comment by AltGr** [comment]
Indeed, opam currently only caches archives, which is the easier case, by their hashes. I considered locally caching VCS repositories too to a common dir below ~/.opam, to avoid multiple fetches when the same VC-controlled URL is needed for several packages or several switches (after all, updating a single clone should be enough to even get different branches). It seemed a little too complex for what it's worth right now so it was delayed, though.

Your case is a bit different since you would want static caching, though. The issue I can see is that opam has no means to tell that `v1.0.0` is a tag and not a branch in your example, so seeing this URL it assumes querying the upstream is always needed. In other words, you don't want to generate archives by default: this might be an option to tell opam to fetch the gits, generate the tarballs, cache them _and_ rewrite the opam/url files to point to the archives (or to otherwise make sure the repository cache is going to be used).

**Comment by hcarty** [comment]
Thank you for the explanation.  A potential implementation could add a `--include-vcs-snapshots` option to `opam admin`. That way the ability is present but the user is required to be explicit about their intent to snapshot a potential changing set of sources.

**Comment by AltGr** [comment]
(Just adding that opam now has a cache of git objects, in reference to my first paragraph above. Nothing to snapshot VCed repositories yet, though)
